### PR TITLE
chore: copy Constructor in 'shared' lib

### PR DESCRIPTION
**Pull request by yeze322**
## Description

Implement the copy constructor in dialogFactory to enable action deep clone.

**Obi types need sepcial treatment**
1. Nested
    - IfCondition
    - SwitchCondition
    - Foreach
    - Foreachpage
    - EditActions
2. With LG templates
    - SendActivity
    - All kinds of *Input (TextInput, ChoiceInput, .etc.)

## Task Item

refs #1193, #1191 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code refactor (non-breaking change which improve code quality, clean up, add tests, etc)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Doc update (document update)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have functionally tested my change

## Screenshots

Please include screenshots or gifs if your PR include UX changes.

### CODE REVIEW

**Comment by yeze322** [comment]
@a-b-r-o-w-n @boydc2014 please review the current pattern we use.

File: Composer/packages/lib/shared/src/types/sdk.ts [hunk 1]
```diff
@@ -3,15 +3,21 @@
...
 /* eslint-disable @typescript-eslint/no-empty-interface */
 
-interface BaseSchema {
+export interface DesignerData {
+  name?: string;
+  description?: string;
+  id: string;
+}
+
+export interface BaseSchema {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/types/sdk.ts]: This was intentionally a private interface. Why do we need to export?

File: Composer/packages/lib/shared/src/types/sdk.ts [hunk 2]
```diff
@@ -3,15 +3,21 @@
...
   /** Defines the valid properties for the component you are configuring (from a dialog .schema file) */
   $type: string;
   /** Inline id for reuse of an inline definition */
   $id?: string;
   /** Copy the definition by id from a .dialog file. */
   $copy?: string;
   /** Extra information for the Bot Framework Composer. */
-  $designer?: OpenObject;
+  $designer?: DesignerData;
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/types/sdk.ts]: 👍

File: Composer/packages/lib/shared/src/types/sdk.ts [hunk 3]
```diff
@@ -229,6 +240,32 @@ export interface SwitchCondition extends BaseSchema {
...
+  itemsProperty?: string;
+  pageSize?: number;
+  actions?: MicrosoftIDialog[];
+}
+
+export interface EditActions extends BaseSchema {
+  changeType: string;
+  actions?: MicrosoftIDialog[];
+}
+
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/types/sdk.ts]: Thanks for improving these!

They should be added to the `MicrosoftIDiloag` union type as well.

File: Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { BaseSchema } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+
+export function shallowCopyAdaptiveAction(input: BaseSchema, externalApi: ExternalApi): BaseSchema {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]:
```suggestion
export function shallowCopyAdaptiveAction<T = MicrosoftIDialog>(input: T, externalApi: ExternalApi): T {
```

File: Composer/packages/lib/shared/src/copyUtils/copySwitchCondition.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import { ExternalApi } from './ExternalApi';
+import { copyAdaptiveActionList } from './copyAdaptiveActionList';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+
+export const copySwitchCondition = async (
+  input: SwitchCondition,
+  externalApi: ExternalApi
+): Promise<SwitchCondition> => {
+  const copy: SwitchCondition = shallowCopyAdaptiveAction(input, externalApi);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copySwitchCondition.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copySendActivity.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { SendActivity } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+
+export const copySendActivity = async (input: SendActivity, externalApi: ExternalApi): Promise<SendActivity> => {
+  const copy: SendActivity = shallowCopyAdaptiveAction(input, externalApi);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copySendActivity.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { InputDialog } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+
+export const copyInputDialog = async (input: InputDialog, externalApi: ExternalApi): Promise<InputDialog> => {
+  const copy: InputDialog = shallowCopyAdaptiveAction(input, externalApi) as InputDialog;
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copyIfCondition.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+// Licensed under the MIT License.
+
+import { IfCondition } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { copyAdaptiveActionList } from './copyAdaptiveActionList';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+
+export const copyIfCondition = async (input: IfCondition, externalApi: ExternalApi): Promise<IfCondition> => {
+  const copy: IfCondition = shallowCopyAdaptiveAction(input, externalApi);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyIfCondition.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copyForeach.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+import { Foreach, ForeachPage } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+import { copyAdaptiveActionList } from './copyAdaptiveActionList';
+
+type ForeachAction = Foreach | ForeachPage;
+export const copyForeach = async (input: ForeachAction, externalApi: ExternalApi): Promise<ForeachAction> => {
+  const copy: ForeachAction = shallowCopyAdaptiveAction(input, externalApi);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyForeach.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copyEditActions.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+// Licensed under the MIT License.
+
+import { EditActions } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+import { copyAdaptiveActionList } from './copyAdaptiveActionList';
+
+export const copyEditActions = async (input: EditActions, externalApi: ExternalApi): Promise<EditActions> => {
+  const copy: EditActions = shallowCopyAdaptiveAction(input, externalApi) as EditActions;
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyEditActions.ts]:
```suggestion
  const copy = shallowCopyAdaptiveAction(input, externalApi);
```

File: Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { copyAdaptiveAction } from './copyAdaptiveAction';
+import { ExternalApi } from './ExternalApi';
+
+export async function copyAdaptiveActionList(actions, externalApi: ExternalApi): Promise<any[]> {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts]:
```suggestion
export async function copyAdaptiveActionList(actions: MicrosoftIDialog[], externalApi: ExternalApi): Promise<MicrosoftIDialog[]> {
```

File: Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  if (!Array.isArray(actions)) return [];
+
+  const results: any[] = [];
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts]:
```suggestion
  const results: MicrosoftIDialog[] = [];
```

File: Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts (hunk 4 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: Template here is great! I will merge all following change requests to one commit instead of accept each of them

File: Composer/packages/lib/shared/src/types/sdk.ts (hunk 1 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/types/sdk.ts]: Not necessariIy. It's a leak, removed.

File: Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts (hunk 4 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: We can either write a generic function here
```ts
function shallowCopy<T extends BaseSchema>(input: T): T {}
```
or use the `MicrosoftIDialog` type directly
```ts
function shallowCopy(input: MicrosoftIDialog): MicrosoftIDialog {}
```
`T = MicrosoftIDialog` seems not working since `MicrosoftIDialog` is already a set of types.
Now I use the second solution to avoid exporting `BaseSchema`.

Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: `MicrosoftIDialog` is a union of types, we cannot use it in a generic function.
image

Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: There is another thing confused me. Semantically, schema ought have a `MicrosoftIAction` interface. In current botbuilder-dotnet schema, `actions` is not differentiated and typed with `MicrosoftIDialog` which means that it could be a recognizer / dialog / trigger. How do you understand that @a-b-r-o-w-n ?

File: Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts (hunk 11 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts]: This is resolved. Typed in commit 3bbc01c4.

File: Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts (hunk 12 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/copyAdaptiveActionList.ts]: resolved by 5562a51

File: Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts (hunk 4 above)
Comment (a-b-r-o-w-n) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: I like `function shallowCopy<T extends BaseSchema = BaseSchema>(input: T): T {}`

Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: Okay great!

Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/shallowCopyAdaptiveAction.ts]: resolved

File: Composer/packages/lib/shared/src/types/sdk.ts (hunk 1 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/types/sdk.ts]: Mark as resolve after we use template in `shallowCopyAdaptiveAction<T as extends>`

File: Composer/packages/lib/shared/src/types/sdk.ts (hunk 3 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/types/sdk.ts]: they are added, mark as resolved

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+    copyLgTemplate: (id, x) => Promise.resolve(x + '(copy)'),
+  };
+
+  it('should return {} when input is invalid', async () => {
+    expect(await copyAdaptiveAction(null, externalApi)).toEqual({});
+    expect(await copyAdaptiveAction({}, externalApi)).toEqual({});
+    expect(await copyAdaptiveAction({ name: 'hi' }, externalApi)).toEqual({});
+  });
+
+  it('can copy BeginDialog', async () => {
```
[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: yes! this is a really good advide

[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: resolved by 69c376

**Comment by yeze322** [comment]
resolved all Andy's comments. now working on adapting new lg api

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+      for (const $type of registeredUniqueTypes) {
+        await copyAdaptiveAction({ $type }, externalApi);
+        expect(CopyConstructorMap[$type]).toHaveBeenCalledTimes(1);
+      }
+
+      for (const $type of registeredUniqueTypes) {
+        await copyAdaptiveAction({ $type }, externalApi);
+        expect(CopyConstructorMap[$type]).toHaveBeenCalledTimes(2);
+      }
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: I don't get this. What is it testing?

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+describe('copyAdaptiveAction', () => {
+  it('should return {} when input is invalid', async () => {
+    expect(await copyAdaptiveAction('hello', externalApi)).toEqual('hello');
+
+    expect(await copyAdaptiveAction(null as any, externalApi)).toEqual({});
+    expect(await copyAdaptiveAction({} as any, externalApi)).toEqual({});
+    expect(await copyAdaptiveAction({ name: 'hi' } as any, externalApi)).toEqual({});
+  });
+
+  describe('should call certain handler in CopyConstructorMap', () => {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: From a test organization point of view, this is not very descriptive.

here's my suggestion for this file:

```
describe('copyAdaptiveAction')
  it('should return {} when input is invalid')
 
  for type in registeredUniqueTypes
    it(`should invoke correct handler for ${type}')

  it('should invoke default handler for other types')
```

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    let originCopyConstructorMap = {};
+    beforeAll(() => {
+      originCopyConstructorMap = { ...CopyConstructorMap };
+    });
+
+    beforeEach(() => {
+      Object.keys(CopyConstructorMap).forEach(key => {
+        CopyConstructorMap[key] = jest.fn();
+      });
+    });
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: This is better done with jest module mocking.

```ts
jest.mock('../../src/copyUtils/copyAdaptiveAction', () => ({
  'Microsoft.TextInput': jest.fn(),
  ...otherTypes
});

// in a test
expect(CopyConstructorMap['Microsoft.TextInput']).toHaveBeenCalledTimes(1);
```

The path may need to be the exact same one used in the app code. I forget.

File: Composer/packages/server/src/models/bot/botProject.ts [hunk 17]
```diff
@@ -182,7 +182,7 @@ export class BotProject {
         ...mainDialog.content.$designer,
         name,
         description,
-      };
+      } as any;
```
Comment (a-b-r-o-w-n) [inline Composer/packages/server/src/models/bot/botProject.ts]: Why is this needed?

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts (hunk 15 above)
Comment (yeze322) [reply Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: great! I will accept this one

File: Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts (hunk 14 above)
Comment (yeze322) [reply Composer/packages/lib/shared/__tests__/copyUtils/copyAdaptiveAction.test.ts]: just to make sure copiers are called every time. I will remove it because of it's not necessary

File: Composer/packages/server/src/models/bot/botProject.ts (hunk 17 above)
Comment (yeze322) [reply Composer/packages/server/src/models/bot/botProject.ts]: the `id` field is missing in this object, triggered a ts error after we adding the `DesingerData` in sdk type. I just realized that it could be rewritten by leveraging the utility in `dialogFactory`.

Comment (yeze322) [reply Composer/packages/server/src/models/bot/botProject.ts]: after understanding the logic here, I rewrite it like this:
image

File: Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+import { InputDialog } from '../types';
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+
+export const copyInputDialog = async (input: InputDialog, externalApi: ExternalApi): Promise<InputDialog> => {
+  const copy = shallowCopyAdaptiveAction(input, externalApi);
+  const nodeId = copy.$designer ? copy.$designer.id : '';
+
+  if (input.prompt) {
```
Comment (boydc2014) [inline Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts]: input.prompt looks like a requried field. so perhaps we can save some checking there?

File: Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    copy.prompt = await externalApi.copyLgTemplate(input.prompt, nodeId);
+  }
+
+  if (input.unrecognizedPrompt) {
```
Comment (boydc2014) [inline Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts]: Similar things like this "unrecoginzedPrompt". Not sure this is required or not, just want to make sure we leverage the type checking as much as possible. 

Because we control the "constructor", we could be able to enforce more strict type (less optional field). Not for just reducing 1 or 2 line codes, more importantly, have the type enforced at first place, will reduce the possible `state` in our system, means we will reduce the overall complexity of the system.

File: Composer/packages/lib/shared/src/copyUtils/copyForeach.ts [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+
+import { ExternalApi } from './ExternalApi';
+import { shallowCopyAdaptiveAction } from './shallowCopyAdaptiveAction';
+import { copyAdaptiveActionList } from './copyAdaptiveActionList';
+
+type ForeachAction = Foreach | ForeachPage;
+export const copyForeach = async (input: ForeachAction, externalApi: ExternalApi): Promise<ForeachAction> => {
+  const copy = shallowCopyAdaptiveAction(input, externalApi);
+
+  if (Array.isArray(input.actions)) {
```
Comment (boydc2014) [inline Composer/packages/lib/shared/src/copyUtils/copyForeach.ts]: is it possible to remove this checking, since you already checked that inside copyList function

**Comment by boydc2014** [comment]
Looks good, i leave a few comments about if we can remove some checking, by enforcing the structure more strictly, which would help reduce the overall complexity.  It might not be the major focus of this PR, but it's a good direction to follow.

File: Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts (hunk 18 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts]: it's declared as required in sdk but might be empty in real dialog json. The checking here is to avoid the case that we generate a json like
```
{
  ...
  prompt: null | undefined
}
```

Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/copyInputDialog.ts]: So the check should not be `if (input.prompt)`, it will miss copying an empty string. Maybe `if (input.prompt !== undefined)`?

File: Composer/packages/lib/shared/src/copyUtils/copyForeach.ts (hunk 20 above)
Comment (yeze322) [reply Composer/packages/lib/shared/src/copyUtils/copyForeach.ts]: it's similar to previous comment, `actions` are declared as optional in SDK. If we don't check it here, the copy result will be different from input value (`[]` against `undefined`)

**Comment by yeze322** [comment]
I got Dong's point, we have declared some fields as required in sdk but during construction we missed the default values for some required fields. Though the 'copy constructor' should respect the original input to copy the data as it was, it's still a 'constructor' which means that it should generate default values for those required fields.

**Comment by yeze322** [comment]
However, I know a strange behavior which makes me cannot assgin a default value for those propmpt fields:
empty-string-in-activity
@boydc2014

**Comment by cwhitten** [comment]
> I got Dong's point, we have declared some fields as required in sdk but during construction we missed the default values for some required fields. Though the 'copy constructor' should respect the original input to copy the data as it was, it's still a 'constructor' which means that it should generate default values for those required fields.

`assignDefaults` exported from `dialogFactory.ts` should handle assigning any/all default values as defined off of the SDK.

**Comment by yeze322** [comment]
@cwhitten I've added default fields for `SendActivity` and `*Input` type. (`initialDialogShape` in `dialogFactory`)
